pointcloud/basic_octree_utils.go

```diff
@@ -8,31 +8,37 @@ import (
 )
 
 // Creates a new LeafNodeEmpty.
-func newLeafNodeEmpty() basicOctreeNode {
+func newLeafNodeEmpty(parent *BasicOctree) basicOctreeNode {
 	octNode := basicOctreeNode{
-		children: nil,
-		nodeType: leafNodeEmpty,
-		point:    PointAndData{},
+		children:     nil,
+		root:         parent,
+		nodeType:     leafNodeEmpty,
+		point:        PointAndData{},
+		maxChildProb: -1,
 	}
 	return octNode
 }
 
 // Creates a new InternalNode with specified children nodes.
-func newInternalNode(tree []*BasicOctree) basicOctreeNode {
+func newInternalNode(tree []*BasicOctree, parent *BasicOctree) basicOctreeNode {
 	octNode := basicOctreeNode{
-		children: tree,
-		nodeType: internalNode,
-		point:    PointAndData{},
+		children:     tree,
+		root:         parent,
+		nodeType:     internalNode,
+		point:        PointAndData{},
+		maxChildProb: -1,
 	}
 	return octNode
 }
 
 // Creates a new LeafNodeFilled and stores specified position and data.
-func newLeafNodeFilled(p r3.Vector, d Data) basicOctreeNode {
+func newLeafNodeFilled(p r3.Vector, d Data, parent *BasicOctree) basicOctreeNode {
 	octNode := basicOctreeNode{
-		children: nil,
-		nodeType: leafNodeFilled,
-		point:    PointAndData{P: p, D: d},
+		children:     nil,
+		root:         parent,
+		nodeType:     leafNodeFilled,
+		point:        PointAndData{P: p, D: d},
+		maxChildProb: d.Value(),
```

Comment:
[opt] If you want to use this outside the octree package you will need a getMaxProb() function